### ISSUE TITLE
Duplicate workflow emails sent in load balanced environment

### ISSUE THREAD

**Issue by nickdunn** [engram-design/Workflow#140]
**Description**
When a user submits content for review, the reviewer is receiving two email notifications.

We are running Craft in a Docker container in Azure. The web app is scaled to two App Service instances, both connected to a single Redis and single MySQL instance, so Azure balances the traffic between. 

I've only seen the duplicate emails occur in this production environment.

Both emails always have exactly the same `Date`

Is there any chance that Workflow could be generating these duplicates, or should I be looking elsewhere?

**Additional info**
- Plugin version: 1.6.4
- Craft version: Craft Pro 3.7.19
- Multi-site: No

**Comment by engram-design** [comment]
Are you able to look in the `workflow_submissions` database table to see if there are duplicate submissions? Just to verify whether this is just an email issue, or if there are multiple submissions being created (which in turn trigger emails).

I don't _think_ these emails should be sending out multiple times, but I'm largely unfamiliar with load balancers and Craft unfortunately. But I don't think that would have any bearing on how emails are sent out, unless code is being run at the same time on multiple instances.

**Comment by nickdunn** [comment]
Yes, they seem to be duplicated in the submissions table. Here's a screenshot with the first instance highlighted.

![image](https://user-images.githubusercontent.com/39560/141017222-687d8fb8-ff2a-4050-b5d5-6dfd769e8949.png)

As you can see, they always seem to be created twice, the first with a null owner and the second a few milliseconds later.

My first hunch was this was someone double-clicking on submit buttons, but it's too regular and across different editors and has started at a specific date/time.

- It's been running load-balanced for some time, although we've only just started using Workflow on this environment.
- On 3/11 we upgraded Craft 3.7.16 to 3.7.19
- 4/11 the submissions are still fine
- First duplicate I can see is on on 5/11

**Comment by engram-design** [comment]
Hmmm, seems more like the draft changes (in Craft) might be coming into play here if the `ownerId` is null.

I'll have to do some digging again.

**Comment by nickdunn** [comment]
Potentially a difference between Craft 3.7.16 and 3.7.19 then?

**Comment by nickdunn** [comment]
@engram-design sorry to hassle, but are you any closer with this — is there any triage/debugging I can do to help diagnose?

**Comment by engram-design** [comment]
I'm still struggling to setup a reproducible environment for this, sorry. I don't suppose gaining access would be possible? Understand if that's not possible.

Do you only access the control panel from a single instance? I don't think there was anything overly major in Craft 3.7.16 to 3.7.19, but there's been plenty of "fixes" it's difficult to keep track

**Comment by nickdunn** [comment]
I'm afraid it's all locked down with IP ranges so I can't get you access. We access the CP via all instances, same as the frontend, Azure is load balancing App Service (https://azure.microsoft.com/en-gb/services/app-service/) instances for us. 

In the same environment I'm also struggling to recreate it, even when logged in as the users submitting content, so I don't know where to go next. The issue persists with these users.

**Comment by engram-design** [comment]
Yep, understandable. So that's interesting that you can't replicate it yourself? I assume your account was part of the editors group, and you made a submission - but just the single submission was created?

~~I'm wondering if you might be able to see what (other) permissions your editor (userId `14778`) has? Can you also try logging into their account and submitting something to Workflow?~~

Edit: Just realised you said you did that 😅

Might it be down to browsers? I also have a feeling it might be due to the auto-save behaviour, but that doesn't make sense as the `ownerId` would still reflect the ID of the entry being autosaved.

Normally when testing the autosave behaviour, you'll need to make a change to content, else Craft won't do anything, so ensure you're changing some form of content on each test.

**Comment by nickdunn** [comment]
I've gone into our UAT environment which _isn't_ load balanced and can see some duplicate rows there too, note the very close created dates. So I don't think it's related to load balancing in our production system.

(image removed for privacy)

Picking the first of these duplicates I can see this draftId in the server logs, seeing the user's activity on the edit entry form through to the final POST which presumably was the final submission to workflow.

(image removed for privacy)

And again include _all_ requests with that draft which shows some auto save AJAX events and the final POST:

(image removed for privacy)

Oddly enough in the UAT environment the `ownerId` is set for these duplicates, not `NULL`. The `data` content is identical when I diff them.

The user agent string suggests Chrome 95/Windows which is very almost the latest version so shouldn't be an issue.

From this example it does look like the duplicate was created from that final save POST request.

**Comment by nickdunn** [comment]
All I can think of is that Craft is publishing two `Entry::EVENT_AFTER_SAVE` events that Workflow subscribes to.

**Comment by engram-design** [comment]
Thanks for the investigation here, I'm frustrated I'm unable to reproduce this. I know there's now a `firstSave` property on the event, which might have something to do with this https://github.com/craftcms/cms/discussions/9638 - but I just can't replicate it to test properly.

I'll have a test now.

**Comment by engram-design** [comment]
Are there any other plugins worth mentioning that your site might be using? Just thinking if something else is triggering the event that isn't on my install.

**Comment by engram-design** [comment]
Wild stab in the dark, but I don't suppose you have a Preparse field in the entry? There's often issues with that plugin, and just to test I've installed it, and that does duplicate saves - but that might be unrelated if you're not using it.

**Comment by engram-design** [comment]
Do I've added a case to handle Preparse fields, and in doing so, I think I might've fixed your issue - although not any further as to _why_ it might be happening. Workflow will now check if the `Entry::EVENT_AFTER_SAVE` event is being called multiple times on the same request, and prevent it from running twice. 

This addresses the Preparse issue, but could likely address your issue.

To get the fix early, change your `verbb/workflow` requirement in `composer.json` to:

```
"require": {
  "verbb/workflow": "dev-craft-3 as 1.6.5",
  "...": "..."
}
```

Then run `composer update`.

**Comment by nickdunn** [comment]
> Wild stab in the dark, but I don't suppose you have a Preparse field in the entry? There's often issues with that plugin, and just to test I've installed it, and that does duplicate saves - but that might be unrelated if you're not using it.

I am, good shout! Only in one section though, so I'll see if I can recreate the issue in that one section only first before pulling your latest version.

**Comment by nickdunn** [comment]
With the original 1.6.4 I was able to recreate the duplicate issue locally with the section that contained the Preparse field. No duplicates from other sections; so this certainly seems like the cause. 

After switching to the updated `dev-craft-3 as 1.6.5` version the issue is fixed :-)

**Comment by nickdunn** [comment]
Thank you for digging into this, and well done for working out it was Preparse!

**Comment by engram-design** [comment]
Aha, bloody Preparse strikes again! I know if there's any element-saving weirdness, that's usually the culprit.

**Comment by nickdunn** [comment]
@engram-design would it be possible to release a new version of Workflow so I can remove the `dev-craft-3 as ...` workaround?

**Comment by engram-design** [comment]
Fixed in [1.6.6](https://github.com/verbb/workflow/releases/tag/1.6.6)
